source/io/module_javaScriptInvoke.js

```diff
@@ -423,35 +435,43 @@ var assignJavaScriptInvoke;
                 invocationState: completionCallbackInvocationEnum.PENDING
             };
 
-            var returnValue = undefined;
             var api;
             if (functionToCall.length === parameters.length + 1) {
                 api = generateAPI(occurrencePointer, functionName, returnValueRef, errorValueRef, completionCallbackStatus, isInternalFunction);
                 parameters.push(api);
             }
 
+            var returnValue;
             try {
                 returnValue = functionToCall.apply(context, parameters);
             } catch (ex) {
-                if (isInternalFunction) {
-                    throw ex;
-                }
+                returnValue = ex;
+            }
 
-                mergeNewError(errorValueRef, functionName, ERRORS.kNIUnableToInvokeAJavaScriptFunction, ex);
-                completionCallbackStatus.retrievalState = completionCallbackRetrievalEnum.UNRETRIEVABLE;
-                completionCallbackStatus.invocationState = completionCallbackInvocationEnum.REJECTED;
+            if (hasExecutionError(returnValue)) {
+                reportExecutionError(functionName, returnValue, errorValueRef, completionCallbackStatus, isInternalFunction);
                 Module.eggShell.setOccurrence(occurrencePointer);
                 return;
             }
 
-            // assume synchronous invocation since the completion callback was never retrieved
+            // assume synchronous invocation since the completion callback was never retrieved by the user
             if (completionCallbackStatus.retrievalState === completionCallbackRetrievalEnum.AVAILABLE) {
-                if (!isInternalFunction) {
-                    tryUpdateReturnValue(functionName, returnValueRef, returnValue, errorValueRef, completionCallbackStatus);
+                updateReturnValue(functionName, returnValueRef, returnValue, errorValueRef, completionCallbackStatus, isInternalFunction);
+                Module.eggShell.setOccurrence(occurrencePointer);
+                return;
+            }
+
+            // for async execution check that user did not provide a return value
+            if (returnValue !== undefined) {
+                if (isInternalFunction) {
+                    throw new Error('Unexpected return value for function requiring asynchronous completion');
                 }
+                mergeNewError(errorValueRef, functionName, ERRORS.kNIUnableToSetReturnValueInJavaScriptInvoke);
```

Comment:
This error happens if a user retrieves the getCompletionCallback() but also gives a return value. We error because the user should not give a return value after getCompletionCallback is called, they should be passing the return value to the getCompletionCallback.

Previously this was silently ignored but now we give the kNIUnableToSetReturnValueInJavaScriptInvoke error, aka:
"Unable to set return value for JavaScript Library Interface node parameter". Should we make a new error code for this scenario or is the unable to set return value error clear enough?

Reply:
I would prefer a new error code. That way we can give precise guidance as to how to fix it or we can use the current one, but we need to update the documentation.

Reply:
That's fair, I'll make that in a separate change.